### ISSUE TITLE
Support tree-shaking unused routes via staticRoutes option

### ISSUE THREAD

**Issue by jelhan** [embroider-build/embroider#822]
Tree-shaking routes, which are not used, would be helpful to use `@embroider/macros` for feature flags. It should tree-shake all routes, which are not registered in `app/router.js`. For backward-compatibility this would need a `staticRoutes` configuration option.

An example use case would look like this:

```js
// app/router.js

import EmberRouter from '@ember/routing/router';
import config from 'my-app/config/environment';
import { getOwnConfig, macroCondition } from '@embroider/macros';

export default class Router extends EmberRouter {
  location = config.locationType;
  rootURL = config.rootURL;
}

Router.map(function () {
  if (macroCondition(getOwnConfig().featureFoo)) {
    this.route('route-introduced-by-feature-foo');
  }
});
```

If `featureFoo` is `false` and `staticRoutes` configuration option is `true` the `route-introduced-by-feature-foo` route, controller and route's template as well as all code only imported by them should not be included in the compiled bundle.

Code splitting at route level could be used as a work-a-round right now. It still produces a chunk for the route but as that route is not used anywhere, that chunk should never be loaded.

To be fully usable for feature flags tree-shaking of modifiers would also be needed. This is tracked in #603.

**Comment by ef4** [comment]
I think this will want to wait for a newer router API. The existing API is not statically analyzable. 

But also, as long as you're splitting out the new route it won't impose any runtime cost, and as long as you're using a macroCondition to guard the router map like in your example, nobody will be able to enter the route with the flag disabled.

**Comment by jelhan** [comment]
Thanks for your quick reply. 

> I think this will want to wait for a newer router API. The existing API is not statically analyzable.

Isn't it statically analyzeable if we put two constraints:

1. All routes must be registered in `app/router.js` using `this.route()`.
2. The route name passed to `this.route()` as first argument must be a static string.

By setting `staticRoutes` to `true` a developer would opt-in to that limitations.

Or would it be still to dynamic?

I agree that code splitting at route-level is a great work-a-round for now. So not a high priority.

**Comment by ef4** [comment]
> Isn't it statically analyzeable if we put two constraints:

No, quite a lot of other constraints would be necessary to do this reliably. For example:

 - must only invoke `map` in an expression statement in the topmost scope
 - must invoke `map` on an identifier that is also exported as default (and there are multiple ways this could be done)
 - must use only a function expression as final argument to `route`, at each depth in the tree
 - must not pass `this` as an argument to any other expression
 - must pass only a string literal to `this.mount()` (and impose all our static constraints on each engine's router map as well)

All of this would interact with the macro system in non-obvious ways, since both these checks and the macros are implemented as babel plugins, which run interleaved with each other at each level of the AST.
